Applications/spire/source/ui/dropdown_menu_list.cpp

```diff
@@ -0,0 +1,184 @@
+  m_scroll_area = new QScrollArea(this);
+  m_scroll_area->setWidgetResizable(true);
+  m_scroll_area->setObjectName("dropdown_menu_list_scroll_area");
+  m_scroll_area->setFrameShape(QFrame::NoFrame);
+  m_scroll_area->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+  m_scroll_area->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+  m_scroll_area->setStyleSheet(QString(R"(
+    #dropdown_menu_list_scroll_area {
+      background-color: #FFFFFF;
+      border-bottom: %2px solid #A0A0A0;
+      border-left: %2px solid #A0A0A0;
+      border-right: %2px solid #A0A0A0;
+      border-top: none;
+    }
+    
+    QScrollBar {
+      background-color: #FFFFFF;
+      border: none;
+    }
+
+    QScrollBar::handle:vertical {
+      background-color: #EBEBEB;
+      margin-left: %3px;
+      width: %1px;
+    }
+
+    QScrollBar::sub-line:vertical {
+      background: none;
+      border: none;
+    }
+
+    QScrollBar::add-line:vertical {
+      background: none;
+      border: none;
+    })").arg(scale_width(15)).arg(scale_width(1)).arg(scale_width(2)));
+  layout->addWidget(m_scroll_area);
+  m_list_widget = new QWidget(m_scroll_area);
+  auto list_layout = new QVBoxLayout(m_list_widget);
+  list_layout->setContentsMargins({});
+  list_layout->setSpacing(0);
+  for(auto& item : items) {
+    auto i = new DropdownMenuItem(item, m_list_widget);
+    i->connect_selected_signal([=] (auto& t) { on_select(t); });
+    list_layout->addWidget(i);
+  }
+  m_list_widget->setStyleSheet("background-color: #FFFFFF;");
+  m_scroll_area->setWidget(m_list_widget);
+  parent->installEventFilter(this);
+}
+
+void DropdownMenuList::set_items(const std::vector<QString>& items) {
+  while(auto item = m_list_widget->layout()->takeAt(0)) {
+    delete item->widget();
+    delete item;
+  }
+  for(auto& item : items) {
+    auto i = new DropdownMenuItem(item, m_list_widget);
+    i->connect_selected_signal([=] (auto& t) { on_select(t); });
+    m_list_widget->layout()->addWidget(i);
+  }
+}
+
+QString DropdownMenuList::get_next(const QString& text) {
+  auto num_items = m_list_widget->layout()->count();
+  auto index = (get_index(text) + 1) % num_items;
+  return static_cast<DropdownMenuItem*>(
+    m_list_widget->layout()->itemAt(index)->widget())->text();
+}
+
+QString DropdownMenuList::get_previous(const QString& text) {
+  auto num_items = m_list_widget->layout()->count();
+  auto index = (get_index(text) - 1) % num_items;
+  if(index < 0) {
+    index = num_items - 1;
+  }
+  return static_cast<DropdownMenuItem*>(
+    m_list_widget->layout()->itemAt(index)->widget())->text();
+}
+
+connection DropdownMenuList::connect_selected_signal(
+    const SelectedSignal::slot_type& slot) const {
+  return m_selected_signal.connect(slot);
+}
+
+bool DropdownMenuList::eventFilter(QObject* object, QEvent* event) {
+  if(object == parent()) {
+    if(event->type() == QEvent::KeyPress && isVisible()) {
+      auto e = static_cast<QKeyEvent*>(event);
```

Comment:
It's best to add precision to your names when resolving a type as opposed to removing precision. That is, when you've resolved that ```event``` is a ```QKeyEvent```, then you should rename that variable to ```key_event``` to make it more specific, rather than making it more vague by naming it ```e```.